src/io/iohk/scalanet/peergroup/DTLSPeerGroup.scala

```diff
@@ -0,0 +1,242 @@
+package io.iohk.scalanet.peergroup
+
+import java.net.InetSocketAddress
+import java.nio.ByteBuffer
+import java.security.cert.Certificate
+import java.security.{PrivateKey, PublicKey}
+import java.util.concurrent.ConcurrentHashMap
+
+import io.iohk.decco.{BufferInstantiator, Codec}
+import io.iohk.scalanet.peergroup.DTLSPeerGroup.Config
+import monix.eval.Task
+import monix.execution.Scheduler
+import monix.reactive.Observable
+import monix.reactive.subjects.{PublishSubject, ReplaySubject, Subject}
+import org.eclipse.californium.elements._
+import org.eclipse.californium.scandium.DTLSConnector
+import org.eclipse.californium.scandium.config.DtlsConnectorConfig
+import org.eclipse.californium.scandium.dtls.cipher.CipherSuite._
+
+import scala.collection.JavaConverters._
+import scala.concurrent.Promise
+
+class DTLSPeerGroup[M](val config: Config)(
+    implicit codec: Codec[M],
+    bufferInstantiator: BufferInstantiator[ByteBuffer],
+    scheduler: Scheduler
+) extends PeerGroup[InetMultiAddress, M] {
+
+  private val serverConnector = createServerConnector()
+
+  private val channelSubject = PublishSubject[Channel[InetMultiAddress, M]]()
+
+  private val activeChannels = new ConcurrentHashMap[(InetSocketAddress, InetSocketAddress), ChannelImpl]().asScala
+
+  override def processAddress: InetMultiAddress = config.processAddress
+
+  override def initialize(): Task[Unit] = {
+    Task(serverConnector.start())
+  }
+
+  override def client(to: InetMultiAddress): Task[Channel[InetMultiAddress, M]] = {
+    val connector = createClientConnector()
+    connector.start()
+    val id = (connector.getAddress, to.inetSocketAddress)
+    assert(!activeChannels.contains(id), s"HOUSTON, WE HAVE A MULTIPLEXING PROBLEM")
+    val channel = new ClientChannelImpl(to, connector)
+    activeChannels.put(id, channel)
+    Task(channel)
+  }
+
+  override def server(): Observable[Channel[InetMultiAddress, M]] = channelSubject
+
+  override def shutdown(): Task[Unit] =
+    for {
+      _ <- Task(serverConnector.stop())
+      _ <- Task(serverConnector.destroy())
+    } yield ()
+
+  private class ChannelImpl(val to: InetMultiAddress, dtlsConnector: DTLSConnector)(implicit codec: Codec[M])
+      extends Channel[InetMultiAddress, M] {
+
+    override val in: Subject[M, M] = ReplaySubject[M]()
```

Comment:
Are you aware that anyone consuming this stream will get all events from the beginning instead of the ones after subscribing, right? otherwise, it can cause weird behavior.